services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java

```diff
@@ -115,57 +162,53 @@ public PortStateJson createPortState(String projectId, PortStateJson portStateJs
         portState.setProjectId(projectId);
 
         try {
-            //Verify VPC ID
-            VpcManagerProxy vpcManagerProxy = new VpcManagerProxy(rollbacks);
-            executor.runAsync(vpcManagerProxy::verifyVpc, portState);
+            createPortStateAsync(portState, executor, rollbacks);
 
-            IpManagerProxy ipManagerProxy = new IpManagerProxy(rollbacks, portState.getProjectId());
-            if (portState.getFixedIps() == null) {
-                executor.runAsync(ipManagerProxy::allocateRandomIpAddress, portState);
-            } else {
-                executor.runAsync(ipManagerProxy::allocateFixedIpAddress, portState.getFixedIps());
-            }
+            //Wait for all async functions to finish
+            executor.joinAll();
 
-            //Generate uuid for port
-            if (portState.getId() == null) {
-                portState.setId(UUID.randomUUID().toString());
-            }
+            //Persist portState
+            portRepository.addItem(portState);
+        } catch (Exception e) {
+            exceptionHandle(executor, rollbacks, e);
+        }
 
-            MacManagerProxy macManagerProxy = new MacManagerProxy(rollbacks);
-            if (portState.getMacAddress() == null) {
-                executor.runAsync(macManagerProxy::allocateRandomMacAddress, portState);
-            } else {
-                executor.runAsync(macManagerProxy::allocateFixedMacAddress, portState);
-            }
+        LOG.info("Create port state success, projectId: {}, PortStateJson: {}", projectId, portStateJson);
 
-            //Verify security group
+        return portStateJson;
+    }
+
+    /**
+     * Create multiple ports, and call the interfaces of each micro-service according to the
+     * configuration of the port to create various required resources for all ports.
+     * If an exception occurs during the creation of multiple ports, we need to roll back
+     * the resource allocated from each micro-service.
+     * @param projectId Project the port belongs to
+     * @param portStateBulkJson Multiple ports configuration
+     * @return PortStateBulkJson
+     * @throws Exception Various exceptions that may occur during the create process
+     */
+    @Override
+    public PortStateBulkJson createPortStateBulk(String projectId, PortStateBulkJson portStateBulkJson) throws Exception {
+        Stack<PortStateRollback> rollbacks = new Stack<>();
+        AsyncExecutor executor = new AsyncExecutor();
 
-            //Verify Binding Host ID
-            if (portState.getBindingHostId() != null) {
-                NodeManagerProxy nodeManagerProxy = new NodeManagerProxy(rollbacks);
-                nodeManagerProxy.verifyHost(portState.getBindingHostId());
+        try {
+            for (PortState portState: portStateBulkJson.getPortStates()) {
+                portState.setProjectId(projectId);
+                createPortStateAsync(portState, executor, rollbacks);
```

Comment:
By sharing the same rollbacks stack, will one port creation failure cause rollback of other ports in the same request?

Reply:
Yes, I think for batch creation, if a port creation fails, all operations should be rolled back.

Reply:
This is right from customer behavior perspective. On the API reliability side, it may not fragile.

Can we add one optional parameter for the API, e.g. allow_partial with default value false. Then the default value is what we are right now. When it is set to true, we will return something all port states, including created/updated port states and failed port with exception.

Reply:
okay, I know what you mean.

Reply:
What does the responsed data look like when we support allow_partial?

Reply:
Supporting allow_partial under the current rollback framework may be a bit troublesome, because we don't know which operations need to be rolled back when an exception occurs.

Reply:
@xieus Regardless of this question, I think this PR is ready to be reviewed.

Reply:
okay then. I will resume the review.

Reply:
Piaoping ~ Have you completed the implementation of port update? When I tested it before, it seemed that the port update API function was not fully completed.

Reply:
@kevin-zhonghao only ip addresses can be updated currently.

Reply:
Use Issue https://github.com/futurewei-cloud/alcor/issues/228 for tracking purpose.